File: daa/python-codes/job_sequencing.py
class Job:
    def __init__ (self, id, deadline, profit):
        self.id = id
        self.deadline = deadline
        self.profit = profit
    
def job_sequencing(jobs):
    jobs.sort(key = lambda x : x.profit, reverse = True)
    max_deadline = max(job.deadline for job in jobs)
    result = [-1] * max_deadline
    slot = [False] * max_deadline
    
    total_profit = 0
    scheduled_jobs = []
    
    for job in jobs:
        for j in range(min(max_deadline - 1, job.deadline - 1), -1, -1):
            if not slot[j]:
                result[j] = job.id
                slot[j] =True
                scheduled_jobs.append(job.id)
                total_profit += job.profit
                break
    
    return scheduled_jobs, total_profit

def main():
    jobs = [
        Job('a', 2, 100),
        Job('b', 1, 19),
        Job('c', 2, 27),
        Job('d', 1, 25),
        Job('e', 3, 15)
    ]
    
    scheduled_jobs, total_profit = job_sequencing(jobs)
    print("Scheduled Jobs:", scheduled_jobs)
    print("Total Profit:", total_profit)

if __name__ == "__main__":    
    main()

File: daa/java-codes/JobSequencing.java
import java.util.*;

class Job {
    String id;
    int deadline;
    int profit;

    Job(String id, int deadline, int profit) {
        this.id = id;
        this.deadline = deadline;
        this.profit = profit;
    }
}

public class JobSequencing {
    public static void main(String[] args) {
        Job[] jobs = {
            new Job("a", 2, 100),
            new Job("b", 1, 19),
            new Job("c", 2, 27),
            new Job("d", 1, 25),
            new Job("e", 3, 15)
        };

        jobSequencing(jobs);
    }

    private static void jobSequencing(Job[] jobs) {
        Arrays.sort(jobs, (a, b) -> b.profit - a.profit);

        int maxDeadline = 0;
        for (Job job : jobs) {
            if (job.deadline > maxDeadline)
                maxDeadline = job.deadline;
        }

        String[] result = new String[maxDeadline];  
        boolean[] slot = new boolean[maxDeadline];  

        int totalProfit = 0;

        for (Job job : jobs) {
            for (int j = Math.min(maxDeadline, job.deadline) - 1; j >= 0; j--) {
                if (!slot[j]) {
                    slot[j] = true;
                    result[j] = job.id;
                    totalProfit += job.profit;
                    break;
                }
            }
        }

        System.out.print("Scheduled Jobs: ");
        for (String jobId : result) {
            if (jobId != null) System.out.print(jobId + " ");
        }
        System.out.println("\nTotal Profit: " + totalProfit);
    }
}
